packages/backend/src/services/CoderService/CoderService.ts

```diff
@@ -0,0 +1,180 @@
+
+            spaceSlug,
+            version: currentVersion,
+        };
+    }
+
+    async getCharts(user: SessionUser, projectUuid: string) {
+        const project = await this.projectModel.get(projectUuid);
+        if (!project) {
+            throw new NotFoundError(`Project ${projectUuid} not found`);
+        }
+
+        // TODO handle permissions
+        // Filter charts based on user permissions (from private spaces)
+
+        // TODO
+        // We need to get the charts and all the chart config
+        // At the moment we are going to fetch them all in individual queries
+        // But in the future we should fetch them in a single query for optimiziation purposes
+        const chartSummaries = await this.savedChartModel.find({ projectUuid });
+        const chartPromises = chartSummaries.map((chart) =>
+            this.savedChartModel.get(chart.uuid),
+        );
+        const charts = await Promise.all(chartPromises);
+
+        // get all spaces to map  spaceSlug
+        const spaceUuids = charts.map((chart) => chart.spaceUuid);
+        const spaces = await this.spaceModel.find({ spaceUuids });
+        return charts.map((chart) =>
+            CoderService.transformChart(chart, spaces),
+        );
+    }
+
+    async upsertChart(
+        user: SessionUser,
+        projectUuid: string,
+        chartAsCode: ChartAsCode,
+    ) {
+        // TODO handle permissions
+
+        const [chart] = await this.savedChartModel.find({
+            slug: chartAsCode.slug,
+        });
+        if (chart === undefined) {
+            // TODO create space if does not exist
+            console.info(
+                `Creating chart ${chartAsCode.name} on project ${projectUuid}`,
+            );
+
+            const createChart: CreateSavedChart & {
+                updatedByUser: UpdatedByUser;
+                slug: string;
+            } = {
+                ...chartAsCode,
+                updatedByUser: user,
+            };
+            const newChart = await this.savedChartModel.create(
+                projectUuid,
+                user.userUuid,
+                createChart,
+            );
+            const spaces = await this.spaceModel.find({
+                spaceUuids: [newChart.spaceUuid],
+            });
+
+            return {
+                chart: CoderService.transformChart(newChart, spaces),
+                created: true,
+            };
+        }
+        // update
+        console.info(
+            `Updating chart ${chartAsCode.name} on project ${projectUuid}`,
+        );
+
+        // TODO update space or create if does not exist
+        await this.savedChartModel.update(chart.uuid, {
+            name: chartAsCode.name,
+            description: chartAsCode.description,
+        });
+
+        const updatedChart = await this.savedChartModel.createVersion(
+            chart.uuid,
+            chartAsCode,
+            user,
+        );
+        const spaces = await this.spaceModel.find({
```

Comment:
we could simply return `created:boolean` without the chart, to avoid this extra request.